Mathlib/Algebra/Group/WithOne/Defs.lean

```diff
@@ -172,6 +172,14 @@ theorem ne_one_iff_exists {x : WithOne α} : x ≠ 1 ↔ ∃ a : α, ↑a = x :=
 #align with_one.ne_one_iff_exists WithOne.ne_one_iff_exists
 #align with_zero.ne_zero_iff_exists WithZero.ne_zero_iff_exists
 
+@[to_additive]
+theorem eq_coe_of_ne_one {x : WithOne α} (hx : x ≠ 1) : ∃ a : α, ↑a = x :=
+  ne_one_iff_exists.mp hx
+
+@[to_additive]
+theorem eq_one_or_coe {x : WithOne α} : x = 1 ∨ ∃ a : α, ↑a = x :=
+  or_iff_not_imp_left.mpr eq_coe_of_ne_one
```

Comment:
I think this should be handled with `induction x` instead; let me make  PR to make that just work...

Reply:
Ah, in fact #12605 already does this

Reply:
Oh my initial proof was a case split but I figured this was cleaner...

Reply:
#13096 is my more-minimal PR that does a small enough number to be a quick review.

Reply:
> Oh my initial proof was a case split but I figured this was cleaner...

What I mean is, the *caller* should be using `induction`, rather than using this lemma with `rcases`

Reply:
If I'm sitting there with `(hx : x ≠ 0)` and I want `n` such that `x = \u n` and I also want to replace `x` with `\u n` everywhere then surely `rcases eq_coe_of_ne_zero hx.ne' with ⟨n, rfl⟩` is hard to beat. You want me to do `induction` + two lines + `subst` instead??

Reply:
My comment only applies to `eq_one_or_coe` (which is the line I commented on), not `eq_coe_of_ne_zero`. The latter is a good lemma.

Reply:
I see -- thanks. I'll check tomorrow that induction works in my use case.

Reply:
Here's a totally magic proof:
```
instance {α : Type*} [Mul α] [Preorder α] [CovariantClass α α (· * ·) (· < ·)]:
    PosMulStrictMono (WithZero α) where
  elim := by
    intro ⟨x, hx⟩ a b (h : a < b)
    dsimp only
    let (x : α) := x
    match a, b with
    | 0, (b : α) =>
      rw [mul_zero]
      exact WithZero.zero_lt_coe _
    | (a : α), (b : α) =>
      norm_cast at h ⊢
      exact mul_lt_mul_left' (α := α) h x
```

Reply:
Or an ever shorter one:
```lean
instance {α : Type*} [Mul α] [Preorder α] [CovariantClass α α (· * ·) (· < ·)]:
    PosMulStrictMono (WithZero α) where
  elim := @fun
    | ⟨(x : α), hx⟩, 0, (b : α), _ => by
        rw [mul_zero]
        exact WithZero.zero_lt_coe _
    | ⟨(x : α), hx⟩, (a : α), (b : α), h => by
        dsimp only
        norm_cast at h ⊢
        exact mul_lt_mul_left' (α := α) h x
```

Reply:
If you have `x ≠ 0` and you want to lift it to `α`, then you need a `CanLift` instance.